xds/src/test/java/io/grpc/xds/CdsLoadBalancerTest.java

```diff
@@ -46,547 +31,485 @@
 import io.grpc.LoadBalancer.SubchannelPicker;
 import io.grpc.LoadBalancerProvider;
 import io.grpc.LoadBalancerRegistry;
-import io.grpc.NameResolver.ConfigOrError;
+import io.grpc.ManagedChannel;
+import io.grpc.NameResolver;
 import io.grpc.Status;
 import io.grpc.Status.Code;
 import io.grpc.SynchronizationContext;
-import io.grpc.internal.FakeClock;
-import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.internal.ObjectPool;
 import io.grpc.xds.CdsLoadBalancerProvider.CdsConfig;
 import io.grpc.xds.EdsLoadBalancerProvider.EdsConfig;
 import io.grpc.xds.EnvoyServerProtoData.UpstreamTlsContext;
-import io.grpc.xds.XdsClient.ClusterUpdate;
-import io.grpc.xds.XdsClient.ClusterWatcher;
-import io.grpc.xds.XdsClient.RefCountedXdsClientObjectPool;
-import io.grpc.xds.XdsClient.XdsClientFactory;
 import io.grpc.xds.internal.sds.CommonTlsContextTestsUtil;
 import io.grpc.xds.internal.sds.SslContextProvider;
 import io.grpc.xds.internal.sds.TlsContextManager;
-import java.net.InetSocketAddress;
-import java.util.ArrayDeque;
+import java.net.SocketAddress;
 import java.util.ArrayList;
-import java.util.Deque;
-import java.util.HashMap;
+import java.util.Collections;
 import java.util.List;
-import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
-import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
 
 /**
  * Tests for {@link CdsLoadBalancer}.
  */
 @RunWith(JUnit4.class)
 public class CdsLoadBalancerTest {
-
-  private final RefCountedXdsClientObjectPool xdsClientPool = new RefCountedXdsClientObjectPool(
-      new XdsClientFactory() {
+  private static final String AUTHORITY = "api.google.com";
+  private static final String CLUSTER = "cluster-foo.googleapis.com";
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
         @Override
-        XdsClient createXdsClient() {
-          xdsClient = mock(XdsClient.class);
-          return xdsClient;
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new AssertionError(e);
         }
+      });
+  private final List<FakeLoadBalancer> childBalancers = new ArrayList<>();
+  private final FakeXdsClient xdsClient = new FakeXdsClient();
+  @Mock
+  private TlsContextManager tlsContextManager;
+  private LoadBalancer.Helper helper = new FakeLbHelper();
+  private int xdsClientRefs;
+  private ConnectivityState currentState;
+  private SubchannelPicker currentPicker;
+  private CdsLoadBalancer loadBalancer;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    when(tlsContextManager.findOrCreateClientSslContextProvider(any(UpstreamTlsContext.class)))
```

Comment:
Implementing a `TlsContextManager` seems simpler and more readable than the boilerplate of `new Answer()`.

Reply:
Yay, it is more reasonable/correct to use a fake implementation here. But the reason is the stuff we want to test/verify more about the side effect (aka, attaching an EAG attribute) than just the interaction with something called a `TlsContextManager`. Fixed.